### ISSUE TITLE
DDPM scheduler dependency on batch size

### ISSUE THREAD

**Issue by gtebbutt** [huggingface/diffusers#6222]
### Describe the bug

I'm not sure if this is "officially" a bug, but it's unexpected enough to me that I wanted to flag it and check before submitting a PR. 

When using the DDPM scheduler, the variance noise uses a single `randn_tensor` the same shape as `model_output` - this means that even with the same prompt and a generator initialised using the same seed, the output for a given image within a batch will vary based on the overall batch size - and will vary significantly for larger batches. 

In short: *every* image in a batch will change, sometimes significantly, based solely on the size of the batch (including the number of images that come after it). My understanding was that batching is intended just for efficiency and the items in a batch shouldn't affect each other? Or at least that later items in a batch shouldn't impact earlier ones. 

Creating the variance noise based on the shape of a single item within `model_output` and then repeating it along the batch axis instead dramatically reduces this effect. That said, it doesn't eliminate it entirely, so if this isn't intended behaviour and it is worth me opening a PR I'll need to do a little more testing just to make sure there aren't any other shape dependencies in there anywhere. 

### Reproduction

```python
import torch
from diffusers import StableDiffusionXLPipeline, EulerDiscreteScheduler, DDPMScheduler

pipe = StableDiffusionXLPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float32, use_safetensors=True).to("cuda")

prompt = "An astronaut on a motorbike"

pipe.scheduler = EulerDiscreteScheduler.from_config(pipe.scheduler.config)

# First image is the same, regardless of how many come after it
pipe(prompt=prompt, width=1344, height=768, generator=torch.Generator(device="cuda").manual_seed(0), num_images_per_prompt=1).images[0].save("euler-bs-1.png")
pipe(prompt=prompt, width=1344, height=768, generator=torch.Generator(device="cuda").manual_seed(0), num_images_per_prompt=16).images[0].save("euler-bs-16.png")

pipe.scheduler = DDPMScheduler.from_config(pipe.scheduler.config)

# First image differs based on how many others are in the batch
pipe(prompt=prompt, width=1344, height=768, generator=torch.Generator(device="cuda").manual_seed(0), num_images_per_prompt=1).images[0].save("ddpm-bs-1.png")
pipe(prompt=prompt, width=1344, height=768, generator=torch.Generator(device="cuda").manual_seed(0), num_images_per_prompt=16).images[0].save("ddpm-bs-16.png")
```

### System Info

- `diffusers` version: 0.24.0
- Platform: Linux-6.2.0-1019-gcp-x86_64-with-glibc2.35
- Python version: 3.10.13
- PyTorch version (GPU?): 2.1.1+cu121 (True)
- Huggingface_hub version: 0.19.4
- Transformers version: 4.35.2
- Accelerate version: 0.25.0
- xFormers version: not installed
- Using GPU in script?: A100
- Using distributed or parallel set-up in script?: No

### Who can help?

@yiyixuxu @patrickvonplaten

**Comment by sayakpaul** [comment]
Cc: @yiyixuxu here.

**Comment by yiyixuxu** [comment]
hi is this related? https://github.com/huggingface/diffusers/issues/6218

**Comment by gtebbutt** [comment]
@yiyixuxu interesting one! It's similar but I don't think it's directly related - looks like that issue was a result of `prepare_latents` calling the same generator four times, because the list (unintentionally) referenced the same generator repeatedly rather than using four different ones. 

In this case, using a single generator so that `prepare_latents` creates different initial noise for each image in the batch is intentional; the outputs are supposed to be different, but the specific issue here is that it's not just the subsequent images that change - the earlier images in the batch change too, depending how many images come after them. 

`DDPMScheduler` only ever calls the generator once per timestep for any batch size, but it "uses up" a different amount of random numbers on each step depending how big the batch is. When the next step starts, the generator state for image zero is in a different position depending how many numbers were taken from it in the previous step. 

If the timestep call always took `(c h w)` random numbers and then repeated that along the batch axis, the generator state on the next step would be identical regardless of batch size, but currently it takes `(b c h w)` random numbers, meaning the next step's starting state depends on `b`.

**Comment by yiyixuxu** [comment]
oh I see now 
I don't think it is a bug here 

can you provide a meaningful use case where it is important to guarantee the first image of the batch remains the same for the same parameters and random seed regardless of the batch size?

**Comment by yiyixuxu** [comment]
ohh actually I think it might be a bug (this also seem to be related https://github.com/huggingface/diffusers/issues/6233#issuecomment-1863809188) 

if we pass a list of generators, we should be able to reproduce any of the single image output with its corresponding seed 

cc @patrickvonplaten - is this something we should fix?

**Comment by gtebbutt** [comment]
Yeah, the minimal reproduction above is just a toy example but it starts to become an issue in more complex use cases, especially when calling some of the lower level scheduler functions directly. It causes unpredictable results when using pipeline callbacks, or when splitting or combining output from multiple unet calls to optimise VRAM use. 

If task batching is being done automatically based on hardware availability, machine load, number of GPUs, etc., and particularly if large batches then need to be resumed from part-way through (e.g. due to a crash) or re-run to find the source of an issue later the batch size dependency can also get very confusing! 

That said, it's not massively difficult to work around now we've figured out the root cause, finding it was the hard part. The broader question I'd ask is: for batch size `x`, assuming `a < x` should it be safe to assume result `a` within that batch is independent of `x`? I think torch takes that approach to batching, for example, so it'd be helpful to set expectations if diffusers doesn't. Seems like a good one to have documented either way, and then a firm yes or no would inherently answer the "is this a bug" question too.

**Comment by patrickvonplaten** [comment]
Yes it's a bug on our side here I think. The reason is that the DDPM scheduler accepts a `generator` object in the step method: https://github.com/huggingface/diffusers/blob/f0a588b8e2783cea0bd1a6b6d7ac85844d91b268/src/diffusers/schedulers/scheduling_ddpm.py#L404 but we don't pass it here: https://github.com/huggingface/diffusers/blob/f0a588b8e2783cea0bd1a6b6d7ac85844d91b268/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L1236

To fix it we should pass the generator object in case the `step(...)` function accepts `generator` as an input. cc @yiyixuxu

**Comment by yiyixuxu** [comment]
hi @gtebbutt 
looked into this a little bit more  - I actually can't reproduce this and do not think there is a bug in diffusers

here is the script I used 

```python
import torch
from diffusers import StableDiffusionXLPipeline, DDPMScheduler

pipe = StableDiffusionXLPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float32, use_safetensors=True)
prompt = "An astronaut on a motorbike"

pipe.scheduler = DDPMScheduler.from_config(pipe.scheduler.config)
pipe.enable_model_cpu_offload()

# First image differs based on how many others are in the batch
pipe(prompt=prompt, width=1344, height=768, generator=torch.Generator(device="cuda").manual_seed(0), num_images_per_prompt=1).images[0].save("yiyi_test_1_ddpm-bs-1.png")

pipe(prompt=prompt, width=1344, height=768, generator=torch.Generator(device="cuda").manual_seed(0), num_images_per_prompt=2).images[0].save("yiyi_test_1_ddpm-bs-2.png")
```

outputs are identical to me 
![yiyi_test_1_ddpm-bs-1](https://github.com/huggingface/diffusers/assets/12631849/c21dfcb2-ffb5-4462-9ddf-87da3652e503)

![yiyi_test_1_ddpm-bs-2](https://github.com/huggingface/diffusers/assets/12631849/dfe66ee8-8713-4aac-b77e-1e0ed7f3f60e)

note that `randn_tensor` is not dependent on batch size 
i.e. this will return `True`

```python

from diffusers.utils.torch_utils import randn_tensor
import torch

generator = torch.Generator(device="cpu").manual_seed(0)
t1 = randn_tensor((1,5), generator = generator)

generator = torch.Generator(device="cpu").manual_seed(0)
t2 = randn_tensor((3,5), generator = generator)[0]

print(torch.allclose(t1, t2))
```

**Comment by gtebbutt** [comment]
Hi @yiyixuxu, thanks for taking the time to look at this! 

In terms of reproducing it visually, the divergence is proportional to the batch size, so it won't necessarily be noticeable between batches of 1 and 2 (although it may be, depending on prompt, LoRAs, etc.) but it can accumulate to significant differences as the value grows:

```python
import torch
from diffusers import StableDiffusionXLPipeline, DDPMScheduler

pipe = StableDiffusionXLPipeline.from_pretrained("stabilityai/stable-diffusion-xl-base-1.0", torch_dtype=torch.float32, use_safetensors=True)
prompt = "An astronaut on a motorbike"

pipe.scheduler = DDPMScheduler.from_config(pipe.scheduler.config)
pipe.enable_model_cpu_offload()

pipe(prompt=prompt, width=1344, height=768, generator=torch.Generator(device="cuda").manual_seed(0), num_images_per_prompt=1).images[0].save("test_1_ddpm-bs-1.png")
pipe(prompt=prompt, width=1344, height=768, generator=torch.Generator(device="cuda").manual_seed(0), num_images_per_prompt=2).images[0].save("test_1_ddpm-bs-2.png")
pipe(prompt=prompt, width=1344, height=768, generator=torch.Generator(device="cuda").manual_seed(0), num_images_per_prompt=4).images[0].save("test_1_ddpm-bs-4.png")
pipe(prompt=prompt, width=1344, height=768, generator=torch.Generator(device="cuda").manual_seed(0), num_images_per_prompt=8).images[0].save("test_1_ddpm-bs-8.png")
pipe(prompt=prompt, width=1344, height=768, generator=torch.Generator(device="cuda").manual_seed(0), num_images_per_prompt=16).images[0].save("test_1_ddpm-bs-16.png")
```

![test_1_ddpm-bs-1](https://github.com/huggingface/diffusers/assets/5956226/70392eb5-dbf0-4417-9f2c-f05b0eb98df3)
![test_1_ddpm-bs-2](https://github.com/huggingface/diffusers/assets/5956226/aabc9256-3275-440d-b79d-8ee60232d982)
![test_1_ddpm-bs-4](https://github.com/huggingface/diffusers/assets/5956226/4b810d98-542d-48d4-915a-eb2963b2dbdb)
![test_1_ddpm-bs-8](https://github.com/huggingface/diffusers/assets/5956226/39a476b3-46ba-4e55-994f-cf5114d7cbb3)
![test_1_ddpm-bs-16](https://github.com/huggingface/diffusers/assets/5956226/604a91e0-46c5-4851-a275-3b4b9796687f)

Although `randn_tensor` isn't dependent on batch size in itself, the state of the generator when used at the beginning of the *next* timestep does depend on how many values from the generator are "used up" by the batch size of a given tensor in the prior step: 

```python
from diffusers.utils.torch_utils import randn_tensor
import torch

generator_1 = torch.Generator(device="cpu").manual_seed(0)
generator_2 = torch.Generator(device="cpu").manual_seed(0)

for i in range(5):
    t1 = randn_tensor((1,5), generator = generator_1)
    t2 = randn_tensor((3,5), generator = generator_2)[0]
    # Only equal on step zero
    print(f"Step {i}, equal? {torch.allclose(t1, t2)}")

generator_1 = torch.Generator(device="cpu").manual_seed(0)
generator_2 = torch.Generator(device="cpu").manual_seed(0)

for i in range(5):
    t1 = randn_tensor((1,5), generator = generator_1)
    t2 = randn_tensor((1,5), generator = generator_2).repeat(3, 1)[0]
    # Remains equal for all steps
    print(f"Step {i}, equal? {torch.allclose(t1, t2)}")
```

**Comment by yiyixuxu** [comment]
cc @patrickvonplaten again, is this considered a "bug"?

**Comment by patrickvonplaten** [comment]
Yes I think this should be considered a bug because as stated [here](https://github.com/huggingface/diffusers/issues/6222#issuecomment-1869757031) we don't pass the generator into the `scheduler.step(...)` function. We should do this though to ensure 100% reproducibility even with varying batch sizes. This means if I pass a list of generators so that I can reproduce the images when changing the batch size. This then also entails that we need to change this part of the code: https://github.com/huggingface/diffusers/blob/f0a588b8e2783cea0bd1a6b6d7ac85844d91b268/src/diffusers/schedulers/scheduling_ddpm.py#L482 so that the computed variance is different for each image in the batch

**Comment by yiyixuxu** [comment]
hi @gtebbutt, I would imagine we have this issue across all schedulers, though - is it the case? 
I don't have a machine that I can test with batch size 16

**Comment by yiyixuxu** [comment]
@gtebbutt for EulerDiscreteScheduler we also use `randn_tensor` to create noise same shape as `model_output` at each step, no?  

https://github.com/huggingface/diffusers/blob/5dc34713800306f498f8ecff6158fcd3668032be/src/diffusers/schedulers/scheduling_euler_discrete.py#L482

**Comment by yiyixuxu** [comment]
@patrickvonplaten 

I want to confirm what's the expected behavior here - do we expect the output to be same in both examples or just example 2. I think we only see difference in example 1 right now (is it? @gtebbutt can confirm) 

example1
```python

generator1 = torch.Generator(device="cuda").manual_seed(0)
generator2 = torch.Generator(device="cuda").manual_seed(0)

out1_bs1 = pipe(..., generator=generator1, num_images_per_prompt=1).images[0]
out1_bs16 = pipe(..., generator=generator2, num_images_per_prompt=16).images[0]
```

example 2 


```python
generator1 = torch.Generator(device="cuda").manual_seed(0)
generator2 = [torch.Generator(device="cuda").manual_seed(seed) for seed in range(16)]

out1_bs1 = pipe(..., generator=generator1, num_images_per_prompt=1).images[0]
out1_bs16 = pipe(..., generator=generator2, num_images_per_prompt=16).images[0]
```

**Comment by yiyixuxu** [comment]
@patrickvonplaten @gtebbutt 

>  we don't pass the generator into the scheduler.step(...) function

We do pass the `generator` argument to ddpm (or any other scheduler that accept `generator` for its step method)  as `extra_step_kwargs` here https://github.com/huggingface/diffusers/blob/f0a588b8e2783cea0bd1a6b6d7ac85844d91b268/src/diffusers/pipelines/stable_diffusion_xl/pipeline_stable_diffusion_xl.py#L566

**Comment by patrickvonplaten** [comment]
Actually, after having taken a closer look at the problem, I don't think we have a bug because as you say we already correctly pass the `generator` and we also correctly compute the variance noise here [diffusers/src/diffusers/schedulers/scheduling_ddpm.py](https://github.com/huggingface/diffusers/blob/f0a588b8e2783cea0bd1a6b6d7ac85844d91b268/src/diffusers/schedulers/scheduling_ddpm.py#L482) because we accept a list of generators there as well.

So if you run the code as proposed by Yiyi in example 2 here: https://github.com/huggingface/diffusers/issues/6222#issuecomment-1887523440 here you should get more or less exactly the same results. 

There might be minor differences because of different CUDA kernels (also see here: https://github.com/huggingface/diffusers/issues/2659)

Also check this page: https://huggingface.co/docs/diffusers/using-diffusers/reusing_seeds to get a better understanding of how to reproduce the same image with varying batch size.
